README.md
[![Project logo](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/logo.png?raw=true "Geometrize Haxe - recreating images as geometric shapes logo")](https://www.geometrize.co.uk/)

[![Haxelib Version](https://img.shields.io/github/tag/Tw1ddle/geometrize-haxe.svg?style=flat-square&label=haxelib)](https://lib.haxe.org/p/geometrize-haxe)
[![Travis Unit Test Build Status](https://img.shields.io/travis/Tw1ddle/geometrize-haxe-unit-tests.svg?style=flat-square)](https://travis-ci.org/Tw1ddle/geometrize-haxe-unit-tests)
[![License](https://img.shields.io/:license-mit-blue.svg?style=flat-square)](https://github.com/Tw1ddle/geometrize-haxe/blob/master/LICENSE)

Geometrize Haxe is a Haxe library for recreating images with geometric primitives. Run the demo [in your browser](https://www.samcodes.co.uk/project/geometrize-haxe-web/).

Geometrize Haxe is part of [Geometrize](https://www.geometrize.co.uk/), and is based on the [primitive](https://github.com/fogleman/primitive) Go library.

[![Geometrized Cat 500 Triangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/cat.gif?raw=true "Geometrized Cat - 500 Geometric Primitives")](https://www.geometrize.co.uk/)

## Features
* Recreate images as geometric primitives - rectangles, rotated rectangles, triangles, circles, ellipses, rotated ellipses and lines are supported.
* Export generated shape data to JSON.
* Export geometrized images as SVGs.
* All Haxe targets are supported.

## Shape Comparison

The matrix shows typical results for circles, triangles, rotated rectangles, rotated ellipses and all supported shapes at 50, 200 and 500 total shapes:

| -                  | 50 Shapes     | 200 Shapes    | 500 Shapes   |
| ------------------ | ------------- | ------------- | ------------ |
| Circles            | [![50 Circles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_50_circles.png?raw=true)](https://www.geometrize.co.uk/) | [![200 Circles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_200_circles.png?raw=true)](https://www.geometrize.co.uk/) | [![500 Circles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_500_circles.png?raw=true)](https://www.geometrize.co.uk/) |
| Triangles          | [![50 Triangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_50_triangles.png?raw=true)](https://www.geometrize.co.uk/) | [![200 Triangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_200_triangles.png?raw=true)](https://www.geometrize.co.uk/) | [![500 Triangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_500_triangles.png?raw=true)](https://www.geometrize.co.uk/) |
| Rotated Rectangles | [![50 Rotated Rectangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_50_rotated_rectangles.png?raw=true)](https://www.geometrize.co.uk/) | [![200 Rotated Rectangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_200_rotated_rectangles.png?raw=true)](https://www.geometrize.co.uk/) | [![500 Rotated Rectangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_500_rotated_rectangles.png?raw=true)](https://www.geometrize.co.uk/) |
| Rotated Ellipses   | [![50 Rotated Ellipses](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_50_rotated_ellipses.png?raw=true)](https://www.geometrize.co.uk/) | [![200 Rotated Ellipses](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_200_rotated_ellipses.png?raw=true)](https://www.geometrize.co.uk/) | [![500 Rotated Ellipses](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_500_rotated_ellipses.png?raw=true)](https://www.geometrize.co.uk/) |
| All Shapes         | [![50 All Shapes](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_50_all_shapes.png?raw=true)](https://www.geometrize.co.uk/) | [![200 All Shapes](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_200_all_shapes.png?raw=true)](https://www.geometrize.co.uk/) | [![500 All Shapes](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/seagull_500_all_shapes.png?raw=true)](https://www.geometrize.co.uk/) |

## How It Works

A user provides a target image, and the algorithm finds shapes to approximate that image. To identify a good shape, the algorithm generates a large number of random candidate shapes, repeatedly improving the fit of each using a hillclimbing optimization approach, eventually choosing the best-fitting shape. The shapes are added one by one.

## JavaScript Usage

If you are working in JavaScript, check out the [geometrize-js](https://github.com/cancerberoSgx/geometrizejs) API and [npm](https://www.npmjs.com/package/geometrizejs) package.

## Haxe Install

Get the [Haxe](https://haxe.org/) library from [GitHub](https://github.com/Tw1ddle/geometrize-haxe) or through [haxelib](https://lib.haxe.org/p/geometrize-haxe/):

```bash
haxelib install geometrize-haxe
```

## Haxe Usage

* Instantiate an ```ImageRunner```, passing it a ```Bitmap``` target image and a starting background ```Rgba``` color.
* Generate shapes by repeatedly calling ```runner.step(options)```, passing in your ```ImageRunnerOptions```.
* Export the results using the ```export``` methods on ```SvgExporter``` and ```ShapeJsonExporter```.

Refer to the library [documentation](https://tw1ddle.github.io/geometrize-haxe/). Also see the Geometrize Haxe [web demo](https://www.samcodes.co.uk/project/geometrize-haxe-web/) and [code](https://github.com/Tw1ddle/geometrize-haxe-web/), or this HaxeFlixel [example](https://samcodes.co.uk/project/geometrize-haxe-flixel/) and [code](https://github.com/Tw1ddle/geometrize-haxe-demo/).

## Resources

* See the Geometrize [resources](https://resources.geometrize.co.uk/) page.

## Examples And Screenshots

Geometrized public domain [artwork](https://commons.wikimedia.org/wiki/Category:Paintings_by_painter) and [photos](https://www.pexels.com/public-domain-images/):

[![Geometrized Flower 330 Rotated Ellipses](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/flower.png?raw=true "Flower - 330 Rotated Ellipses")](https://www.geometrize.co.uk/)
[![Geometrized Train 230 Rotated Ellipses](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/train.png?raw=true "Train - 230 Rotated Ellipses")](https://www.geometrize.co.uk/)
[![Geometrized Woodland Cemetery 600 Rotated Rectangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/woodland_cemetery.png?raw=true "Woodland Cemetery - 600 Rotated Rectangles")](https://www.geometrize.co.uk/)
[![Geometrized Pomegranate 300 Rotated Ellipses](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/pomegranate.png?raw=true "Pomegranate - 300 Rotated Ellipses")](https://www.geometrize.co.uk/)
[![Geometrized Monarch Butterfly 800 Various Shapes](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/monarch_butterfly.png?raw=true "Monarch Butterfly - 800 Various Shapes")](https://www.geometrize.co.uk/)
[![Geometrized Leafy Railroad 800 Rotated Rectangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/leafy_railroad.png?raw=true "Leafy Railroad - 800 Rotated Rectangles")](https://www.geometrize.co.uk/)
[![Geometrized Pyramid 150 Triangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/pyramid.png?raw=true "Pyramid - 150 Triangles")](https://www.geometrize.co.uk/)
[![Geometrized Trees 210 Ellipses](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/tree_under_clouds.png?raw=true "Tree Under Clouds - 210 Ellipses")](https://www.geometrize.co.uk/)
[![Geometrized Chomsky 300 Triangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/chomsky.png?raw=true "Noam Chomsky - 300 Triangles")](https://www.geometrize.co.uk/)
[![Geometrized Trees 250 Rotated Ellipses](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/trees.png?raw=true "Trees - 250 Rotated Ellipses")](https://www.geometrize.co.uk/)
[![Geometrized Fairies 500 Triangles](https://github.com/Tw1ddle/geometrize-haxe/blob/master/screenshots/fairies.png?raw=true "Fairies - 500 Triangles")](https://www.geometrize.co.uk/)

For more examples, see the Geometrize [gallery](https://gallery.geometrize.co.uk/).

## Notes
* This implementation is single-threaded, and performance varies by target platform. Small target images work best.
* Got an idea or suggestion? Open an issue on GitHub, or send Sam a message on [Twitter](https://twitter.com/Sam_Twidale).
* Find more related projects and open source code [here](https://resources.geometrize.co.uk/).

CHANGELOG.md
## 1.0.0
* Initial release.

## 1.0.1
* Minor tweaks to make image runner settings simpler to pass to a web worker on the js target.
* Made some SVG exporter helper methods public.

## 1.0.2
* Minor tweaks to JSON export functions so individual shapes can be more easily exported to JSON.

## 1.0.3
* Expose bits of the API so other applications written in Javascript can use it.

examples/javascript-node/node-example.js
const { geometrize } = require('../../build/geometrize')
const { writeFileSync, existsSync, readFileSync } = require('fs')
const Jimp = require('jimp')

async function loadPng(url) {
  const image = await  Jimp.read(url)
  return geometrize.bitmap.Bitmap.createFromRawBytes(image.bitmap.width, image.bitmap.height, image.bitmap.data)
}

async function run({ input, iterations, candidateShapesPerStep, shapeMutationsPerStep, alpha, shapeTypes }) {
  try {
    const bitmap = await loadPng(input)
    const runner = new geometrize.runner.ImageRunner(bitmap)
    const options = {
      shapeTypes, // TODO: from enum
      candidateShapesPerStep,
      shapeMutationsPerStep,
      alpha
    }
    const svgData = []
    for (let i = 0; i < iterations; i++) {
      svgData.push(geometrize.exporter.SvgExporter.exportShapes(runner.step(options)))
    }
    return geometrize.exporter.SvgExporter.getSvgPrelude() + geometrize.exporter.SvgExporter.getSvgNodeOpen(bitmap.width, bitmap.height) + svgData.join('\n') + geometrize.exporter.SvgExporter.getSvgNodeClose()
  } catch (error) {
    console.error(`An error ocurred:`, error);
    process.exit(1)
  }
}

async function main() {
  const t0 = Date.now()
  const args = require('minimist')(process.argv.slice(2)) || {}
  const options = {
    ...{
      input: 'docs/logo.png',
      output: 'tmp_out.svg',
      iterations: 1000,
      shapeTypes: [0, 1, 2, 3, 4, 5, 6], // TODO: from enum
      candidateShapesPerStep: 50,
      shapeMutationsPerStep: 100,
      alpha: 128,
    }, ...args
  }
  if(typeof options.shapeTypes==='string' ){
    options.shapeTypes = options.shapeTypes.split(',').map(s=>parseInt(s))
  }

  const svg = await run(options)
  writeFileSync(options.output, svg)
  console.log(`
Written file ${options.output}.
Elapsed time: ${(Date.now() - t0) / 1000} seconds.
Output size: ${svg.length / 1000} KB.
Options: ${JSON.stringify(options)}
`.trim())

}

main()


examples/javascript-browser/browser-example.js
const { geometrize } = require('../../build/geometrize')
const PNGReader = require('png.js')

function loadPng(url) {
  return new Promise((resolve, reject) => {
    fetch(url).then(response =>
      response.arrayBuffer()
    ).then(bytes => {
      const reader = new PNGReader(bytes)
      reader.parse(function (err, png) {
        if (err) reject(err)
        document.getElementById('inputSize').innerHTML = (png.pixels.length / 1000) + 'KB'
        resolve(geometrize.bitmap.Bitmap.createFromRawBytes(png.width, png.height, png.pixels))
      })
    })
  })
}

function renderSvg(bitmap, svgData) {
  const svg = geometrize.exporter.SvgExporter.getSvgPrelude() + geometrize.exporter.SvgExporter.getSvgNodeOpen(bitmap.width, bitmap.height) + svgData.join('\n') + geometrize.exporter.SvgExporter.getSvgNodeClose()
  document.getElementById('output').innerHTML = svg
  document.getElementById('outputSize').innerHTML = (svg.length / 1000) + 'KB'
}

function wait(ms) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve()
    }, ms)
  })
}

async function run({ imgUrl, iterations, interval, candidateShapesPerStep, shapeMutationsPerStep, alpha }) {
  const bitmap = await loadPng(imgUrl)
  const runner = new geometrize.runner.ImageRunner(bitmap)
  const options = {
    shapeTypes: [0, 1, 2, 3, 4, 5, 6], // TODO: from enum
    candidateShapesPerStep,
    shapeMutationsPerStep,
    alpha
  }
  const svgData = []
  for (let i = 0; i < iterations; i++) {
    svgData.push(geometrize.exporter.SvgExporter.exportShapes(runner.step(options)))
    if (interval) {
      renderSvg(bitmap, svgData)
      await wait(interval)
    }
  }
  renderSvg(bitmap, svgData)
}

async function test() {
  document.getElementById('main').innerHTML = html
  document.getElementById('run').addEventListener('click', async e => {
    document.getElementById('status').innerHTML = 'Working'
    const t0 = Date.now()
    await run({
      imgUrl: document.getElementById('imgUrl').value || 'https://upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Fiji_banded_iguana_in_Vienna_Zoo_on_2013-05-12.png/640px-Fiji_banded_iguana_in_Vienna_Zoo_on_2013-05-12.png',
      iterations: parseInt(document.getElementById('iterations').value || '100'),
      interval: parseInt(document.getElementById('interval').value || '0'),
      alpha: parseInt(document.getElementById('alpha').value || '0'),
      candidateShapesPerStep: parseInt(document.getElementById('candidateShapesPerStep').value || '0'),
      shapeMutationsPerStep: parseInt(document.getElementById('shapeMutationsPerStep').value || '0'),
    })
    document.getElementById('elapsedTime').innerHTML = ((Date.now() - t0) / 1000) + 'seconds'
    document.getElementById('status').innerHTML = 'Stand by'
  })
  document.getElementById('imgUrl').addEventListener('change', async e => {
    document.getElementById('img').src = document.getElementById('imgUrl').value
  })
}

const html = `
<img id="img" src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Fiji_banded_iguana_in_Vienna_Zoo_on_2013-05-12.png/640px-Fiji_banded_iguana_in_Vienna_Zoo_on_2013-05-12.png"/><br/>
<div>Input size: <span id="inputSize"></span></div>
<label>Image URL: <input type="url" value="https://upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Fiji_banded_iguana_in_Vienna_Zoo_on_2013-05-12.png/640px-Fiji_banded_iguana_in_Vienna_Zoo_on_2013-05-12.png" id="imgUrl"/></label><br/>
<label>Iterations: <input id="iterations" type="number" min="1" value="100"/></label><br/>
<label>Alpha: <input id="alpha" type="number" min="1" value="128"/></label><br/>
<label>Candidate shapes per step: <input id="candidateShapesPerStep" type="number" min="1" value="50"/></label><br/>
<label>Shape mutation per step: <input id="shapeMutationsPerStep" type="number" min="1" value="100"/></label><br/>

<label>Step interval (0 to prevent render in steps): <input id="interval" type="number" min="0" value="0"/></label><br/>
<button id="run">run</button><br/>
<div>Status: <span id="status">Stand by</span></div>
<div>Output size: <span id="outputSize"></span></div>
<div>Elapsed time: <span id="elapsedTime"></span></div>
<div id="output"></div>
<h4>Working Image URLs</h4>
<p>png.js, the library to decode images that this example uses, support a limited set of formats (this is not a geometrize limitation but png.js'). This is a list of images that currently work for you to test:</p>
<ul>
<li>https://upload.wikimedia.org/wikipedia/commons/4/47/PNG_transparency_demonstration_1.png</li>
<li>logo.png</li>
<li>https://upload.wikimedia.org/wikipedia/commons/b/b8/Lokazzjoni_ta%27_Ruma.png</li>
<li>https://upload.wikimedia.org/wikipedia/commons/1/15/Xochiquetzal_V.png</li>
<li>https://i.imgur.com/PzdDHyo.png?2</li>
<li>https://i.imgur.com/8UQhIQQ.png</li>
<li>https://upload.wikimedia.org/wikipedia/commons/8/86/Nikon_D3000_CN-2011-04.png</li>
<li>https://upload.wikimedia.org/wikipedia/commons/d/d6/Wikipedia-logo-v2-en.png</li>
<li>https://upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Fiji_banded_iguana_in_Vienna_Zoo_on_2013-05-12.png/640px-Fiji_banded_iguana_in_Vienna_Zoo_on_2013-05-12.png</li>
</ul>
`

test()
